File: Gap-Controlled-Simplex/Solvers/IterativeSolver.cs
namespace Gap_Controlled_Simplex.Solvers;

public abstract class IterativeSolver : ISolver
{
    public int? MaxIterations = null;

    protected bool checkIterationCount(int iterations)
        => !MaxIterations.HasValue || iterations <= MaxIterations.Value;

    public abstract Solution? Maximize(in Problem p, int[]? startingBasis);


    public abstract (Vertex? v, int initialIterations) 
        GetStartingPoint(in Problem p, int[]? givenBasis = null);
}

File: src/gsimplex/solvers/iterative_solver.py
from abc import ABC, abstractmethod
from typing import Optional, Tuple, List, Union
from pulp import LpProblem, LpConstraint

from gsimplex.solvers.solver_interface import ISolver
from gsimplex.vertex import Vertex

class IterativeSolver(ISolver, ABC):
    def __init__(self):
        self.max_iterations: Optional[int] = None

    def _check_iteration_count(self, iterations: int) -> bool:
        return self.max_iterations is None or iterations <= self.max_iterations

    @abstractmethod
    def get_starting_point(self, 
                           problem: LpProblem, 
                           given_basis: Optional[Union[List[LpConstraint], Vertex]] = None
                           ) -> Tuple[Optional[Vertex], int]:
        pass